File: core/src/muscle_activation.cpp
#include "workoutlog/muscle_activation.hpp"

#include <algorithm>

#include "workoutlog/catalogue.hpp"

namespace workoutlog {

std::string to_string(WeightingMode m) {
    switch (m) {
        case WeightingMode::set_count: return "set_count";
        case WeightingMode::rep_volume: return "rep_volume";
        case WeightingMode::tonnage: return "tonnage";
    }
    return {};
}

std::string label(WeightingMode m) {
    switch (m) {
        case WeightingMode::set_count: return "Sets";
        case WeightingMode::rep_volume: return "Reps";
        case WeightingMode::tonnage: return "Tonnage";
    }
    return {};
}

namespace {

int reps_of(const WorkSet& s) {
    if (s.reps) return *s.reps;
    if (s.total_reps) return *s.total_reps;
    if (s.cluster) {
        int sum = 0;
        for (int r : *s.cluster) sum += r;
        return sum;
    }
    return 0;
}

double strength_volume(const std::vector<WorkSet>& sets, WeightingMode mode) {
    switch (mode) {
        case WeightingMode::set_count:
            return static_cast<double>(sets.size());
        case WeightingMode::rep_volume: {
            double total = 0;
            for (const auto& s : sets) total += reps_of(s);
            return total;
        }
        case WeightingMode::tonnage: {
            double total = 0;
            for (const auto& s : sets) total += reps_of(s) * s.weight_kg.value_or(0.0);
            return total;
        }
    }
    return 0;
}

// Preserves a known quirk of the Swift original: this reads block.scheme / the
// exercise's reps_override, never MetconRound::reps, so an AMRAP's actual completed
// rounds don't feed the map. Not fixed here deliberately -- see docs/05-architecture.md.
double metcon_volume(const MetconBlock& block, const MetconExercise& exercise, WeightingMode mode) {
    std::vector<int> scheme = exercise.reps_override.value_or(block.scheme.value_or(std::vector<int>{}));
    size_t rounds = block.rounds ? block.rounds->size() : scheme.size();

    switch (mode) {
        case WeightingMode::set_count: {
            size_t s = scheme.empty() ? 1 : scheme.size();
            return static_cast<double>(std::max(rounds, s));
        }
        case WeightingMode::rep_volume: {
            double total = 0;
            for (int r : scheme) total += r;
            return total;
        }
        case WeightingMode::tonnage: {
            double sum = 0;
            for (int r : scheme) sum += r;
            return sum * exercise.weight_kg.value_or(0.0);
        }
    }
    return 0;
}

std::map<std::string, double> normalize(std::map<std::string, double> raw) {
    double peak = 0;
    for (const auto& [k, v] : raw) peak = std::max(peak, v);
    if (peak <= 0) return raw;
    for (auto& [k, v] : raw) v /= peak;
    return raw;
}

} // namespace

std::map<std::string, double> MuscleActivation::for_exercise(const Exercise& exercise) const {
    std::map<std::string, double> raw;
    for (const auto& m : exercise.primary_muscles) raw[m] = std::max(raw[m], primary_weight_);
    for (const auto& m : exercise.secondary_muscles) raw[m] = std::max(raw[m], secondary_weight_);
    return normalize(std::move(raw));
}

void MuscleActivation::accumulate(const Session& session, std::map<std::string, double>& raw,
                                   const Catalogue& catalogue, WeightingMode mode) const {
    auto add = [&](const Exercise& exercise, double volume) {
        if (volume <= 0) return;
        for (const auto& m : exercise.primary_muscles) raw[m] += volume * primary_weight_;
        for (const auto& m : exercise.secondary_muscles) raw[m] += volume * secondary_weight_;
    };

    for (const auto& block : session.blocks) {
        if (const auto* s = std::get_if<StrengthBlock>(&block)) {
            const Exercise* ex = catalogue::resolve(catalogue, s->exercise);
            if (!ex) continue;
            add(*ex, strength_volume(s->sets, mode));
        } else if (const auto* m = std::get_if<MetconBlock>(&block)) {
            for (const auto& me : m->exercises) {
                const Exercise* ex = catalogue::resolve(catalogue, me.name);
                if (!ex) continue;
                add(*ex, metcon_volume(*m, me, mode));
            }
        }
        // cardio, cooldown: no muscle contribution.
    }
}

std::map<std::string, double> MuscleActivation::for_session(const Session& session, const Catalogue& catalogue,
                                                              WeightingMode mode) const {
    std::map<std::string, double> raw;
    accumulate(session, raw, catalogue, mode);
    return normalize(std::move(raw));
}

std::map<std::string, double> MuscleActivation::for_sessions(std::span<const Session> sessions,
                                                               const Catalogue& catalogue, WeightingMode mode) const {
    std::map<std::string, double> raw;
    for (const auto& session : sessions) accumulate(session, raw, catalogue, mode);
    return normalize(std::move(raw));
}

} // namespace workoutlog

File: crates/workout-log-core/src/analytics/muscle_activation.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use crate::models::block::Block;
use crate::models::exercise::{Catalogue, Exercise};
use crate::models::metcon::{MetconBlock, MetconExercise};
use crate::models::session::Session;
use crate::models::work_set::WorkSet;

pub type Scores = BTreeMap<String, f64>;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WeightingMode {
    #[default]
    SetCount,
    RepVolume,
    Tonnage,
}

impl WeightingMode {
    pub const ALL: [Self; 3] = [Self::SetCount, Self::RepVolume, Self::Tonnage];

    pub fn wire(self) -> &'static str {
        match self {
            Self::SetCount => "set_count",
            Self::RepVolume => "rep_volume",
            Self::Tonnage => "tonnage",
        }
    }

    pub fn label(self) -> &'static str {
        match self {
            Self::SetCount => "Sets",
            Self::RepVolume => "Reps",
            Self::Tonnage => "Tonnage",
        }
    }

    pub fn from_wire(wire: &str) -> Option<Self> {
        Self::ALL.into_iter().find(|m| m.wire() == wire)
    }
}

#[derive(Debug, Clone, Copy)]
pub struct MuscleActivation {
    pub primary_weight: f64,
    pub secondary_weight: f64,
}

impl Default for MuscleActivation {
    fn default() -> Self {
        Self {
            primary_weight: 1.0,
            secondary_weight: 0.5,
        }
    }
}

impl MuscleActivation {
    /// A muscle listed as both primary and secondary keeps the primary weight.
    pub fn for_exercise(&self, exercise: &Exercise) -> Scores {
        let mut raw = Scores::new();
        for muscle in &exercise.primary_muscles {
            let slot = raw.entry(muscle.clone()).or_insert(0.0);
            *slot = slot.max(self.primary_weight);
        }
        for muscle in exercise.secondary_muscles() {
            let slot = raw.entry(muscle.clone()).or_insert(0.0);
            *slot = slot.max(self.secondary_weight);
        }
        normalize(raw)
    }

    pub fn for_session(
        &self,
        session: &Session,
        catalogue: &Catalogue,
        mode: WeightingMode,
    ) -> Scores {
        let mut raw = Scores::new();
        self.accumulate(&mut raw, session, catalogue, mode);
        normalize(raw)
    }

    /// Raw volume is summed across sessions and normalized once, rather than
    /// averaging maps that were each already normalized.
    pub fn for_sessions(
        &self,
        sessions: &[Session],
        catalogue: &Catalogue,
        mode: WeightingMode,
    ) -> Scores {
        let mut raw = Scores::new();
        for session in sessions {
            self.accumulate(&mut raw, session, catalogue, mode);
        }
        normalize(raw)
    }

    fn accumulate(
        &self,
        raw: &mut Scores,
        session: &Session,
        catalogue: &Catalogue,
        mode: WeightingMode,
    ) {
        for block in &session.blocks {
            match block {
                Block::Strength(strength) => {
                    if let Some(exercise) = catalogue.resolve(&strength.exercise) {
                        self.add(raw, exercise, strength_volume(&strength.sets, mode));
                    }
                }
                Block::Metcon(metcon) => {
                    for entry in &metcon.exercises {
                        if let Some(exercise) = catalogue.resolve(&entry.name) {
                            self.add(raw, exercise, metcon_volume(metcon, entry, mode));
                        }
                    }
                }
                Block::Cardio(_) | Block::Cooldown(_) => {}
            }
        }
    }

    fn add(&self, raw: &mut Scores, exercise: &Exercise, volume: f64) {
        if volume <= 0.0 {
            return;
        }
        for muscle in &exercise.primary_muscles {
            *raw.entry(muscle.clone()).or_insert(0.0) += volume * self.primary_weight;
        }
        for muscle in exercise.secondary_muscles() {
            *raw.entry(muscle.clone()).or_insert(0.0) += volume * self.secondary_weight;
        }
    }
}

fn strength_volume(sets: &[WorkSet], mode: WeightingMode) -> f64 {
    match mode {
        WeightingMode::SetCount => sets.len() as f64,
        WeightingMode::RepVolume => sets
            .iter()
            .map(|s| s.recorded_reps().unwrap_or(0) as f64)
            .sum(),
        WeightingMode::Tonnage => sets.iter().map(WorkSet::tonnage_kg).sum(),
    }
}

fn metcon_volume(block: &MetconBlock, entry: &MetconExercise, mode: WeightingMode) -> f64 {
    let scheme = entry.scheme_within(block.scheme.as_deref());
    let rounds = block.rounds.as_ref().map_or(scheme.len(), Vec::len);
    let scheme_reps: i64 = scheme.iter().sum();
    match mode {
        WeightingMode::SetCount => {
            let slots = if scheme.is_empty() { 1 } else { scheme.len() };
            rounds.max(slots) as f64
        }
        WeightingMode::RepVolume => scheme_reps as f64,
        WeightingMode::Tonnage => scheme_reps as f64 * entry.weight_kg.unwrap_or(0.0),
    }
}

fn normalize(raw: Scores) -> Scores {
    let peak = raw.values().copied().fold(f64::NEG_INFINITY, f64::max);
    if raw.is_empty() || peak <= 0.0 {
        return raw;
    }
    raw.into_iter().map(|(k, v)| (k, v / peak)).collect()
}
